diff -urN rovio_ros2_ref/src/Camera.cpp reaqrovio_ros1_ref/src/Camera.cpp
--- rovio_ros2_ref/src/Camera.cpp	2026-07-24 19:53:08.746850352 +0300
+++ reaqrovio_ros1_ref/src/Camera.cpp	2026-07-24 19:52:58.992974228 +0300
@@ -1,11 +1,72 @@
+/*
+* For parts of code or files added at the Autonomous Robots Lab, Norwegian University of Science and Technology, the following license is applicable:
+*
+* Copyright (c) 2024, Autonomous Robots Lab, Norwegian University of Science and Technology
+*
+* Redistribution and use in source and binary forms, with or without
+* modification, are permitted provided that the following conditions are met:
+*
+* 1. Redistributions of source code must retain the above copyright notice, this
+*    list of conditions and the following disclaimer.
+*
+* 2. Redistributions in binary form must reproduce the above copyright notice,
+*    this list of conditions and the following disclaimer in the documentation
+*    and/or other materials provided with the distribution.
+*
+* 3. Neither the name of the copyright holder nor the names of its
+*    contributors may be used to endorse or promote products derived from
+*    this software without specific prior written permission.
+*
+* THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+* AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+* IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+* DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+* FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+* DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+* SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+* CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+* OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+* OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*
+* For parts of code or files directly adopted from the Autonomous Systems Lab, the following license is applicable (as from original repository for ROVIO: https://github.com/ethz-asl/rovio):
+*
+* Copyright (c) 2014, Autonomous Systems Lab
+* All rights reserved.
+*
+* Redistribution and use in source and binary forms, with or without
+* modification, are permitted provided that the following conditions are met:
+* * Redistributions of source code must retain the above copyright
+* notice, this list of conditions and the following disclaimer.
+* * Redistributions in binary form must reproduce the above copyright
+* notice, this list of conditions and the following disclaimer in the
+* documentation and/or other materials provided with the distribution.
+* * Neither the name of the Autonomous Systems Lab, ETH Zurich nor the
+* names of its contributors may be used to endorse or promote products
+* derived from this software without specific prior written permission.
+*
+* THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+* ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+* WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+* DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+* FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+* DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+* SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+* CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+* OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+* OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*/
+
 #include "rovio/Camera.hpp"
 #include "yaml-cpp/yaml.h"
+#include <opencv2/imgproc/imgproc.hpp>
+#include <opencv2/calib3d.hpp>
 
 namespace rovio{
 
   Camera::Camera(){
     k1_ = 0.0; k2_ = 0.0; k3_ = 0.0; k4_ = 0.0; k5_ = 0.0; k6_ = 0.0;
     p1_ = 0.0; p2_ = 0.0; s1_ = 0.0; s2_ = 0.0; s3_ = 0.0; s4_ = 0.0;
+    refrac_ind_ = 1.0;
     K_.setIdentity();
     type_ = RADTAN;
     valid_radius_ = std::numeric_limits<double>::max();
@@ -25,20 +86,6 @@
     K_(2,1) = config["camera_matrix"]["data"][7].as<double>();
     K_(2,2) = config["camera_matrix"]["data"][8].as<double>();
     std::cout << "Set Camera Matrix to:\n" << K_ << std::endl;
-    image_width_ = config["image_width"].as<int>();
-    image_height_ = config["image_height"].as<int>();
-  }
-
-  void Camera::scaleCameraMatrix(const double scale_x, const double scale_y) {
-    K_(0,0) = scale_x * K_(0,0);
-    K_(0,1) =  K_(0,1);
-    K_(0,2) = scale_x * K_(0,2);
-    K_(1,0) =  K_(1,0);
-    K_(1,1) = scale_y * K_(1,1);
-    K_(1,2) = scale_y * K_(1,2);
-    K_(2,1) =  K_(2,1);
-    K_(2,2) =  K_(2,2);
-    std::cout << "Set camera matrix to: \n" << K_ << std::endl;
   }
 
   void Camera::loadRadtan(const std::string& filename){
@@ -52,6 +99,14 @@
     std::cout << "Set distortion parameters (Radtan) to: k1(" << k1_ << "), k2(" << k2_ << "), k3(" << k3_ << "), p1(" << p1_ << "), p2(" << p2_ << ")" << std::endl;
   }
 
+  void Camera::loadRefractive(const std::string& filename){
+    loadCameraMatrix(filename);
+    YAML::Node config = YAML::LoadFile(filename);
+    // refrac_ind_ = config["refractive_index"]["data"][0].as<double>();
+
+    std::cout << "Set distortion parameters (Refractive) to: refrac_ind(" << refrac_ind_ << ")" << std::endl;
+  }
+
   void Camera::loadEquidist(const std::string& filename){
     loadCameraMatrix(filename);
     YAML::Node config = YAML::LoadFile(filename);
@@ -82,10 +137,19 @@
     if(distortionModel == "plumb_bob"){
       type_ = RADTAN;
       loadRadtan(filename);
+    } else if(distortionModel == "refractive"){
+      type_ = REFRAC;
+      loadRefractive(filename);
     } else if(distortionModel == "equidistant"){
       type_ = EQUIDIST;
       loadEquidist(filename);
-    } else if(distortionModel == "ds"){
+    } else if(distortionModel == "equirefractive"){
+      std::cout << "Loading equirefractive model" << std::endl;
+      type_ = EQUIREFRAC;
+      loadEquidist(filename);
+      loadRefractive(filename);
+    }
+    else if(distortionModel == "ds"){
       type_ = DS;
       loadDoubleSphere(filename);
     } else {
@@ -117,6 +181,108 @@
     J(1,1) = kr + 2.0 * k1_ * y2 + 4.0 * k2_ * y2 * r2 + 6.0 * k3_ * y2 * r2 * r2 + 6.0 * p1_ * in(1) + 2.0 * p2_ * in(0);
   }
 
+  void Camera::distortRefractive(const Eigen::Vector2d& in, Eigen::Vector2d& out) const{
+    const double x2 = in(0) * in(0);
+    const double y2 = in(1) * in(1);
+    const double x_y = in(0) * in(1);
+    const double r2 = x2 + y2;
+    const double n = refrac_ind_;
+    const double n2 = n * n;
+
+    const double m_distort = n/sqrt(1 + r2 - (n2*r2));
+    out(0) = in(0) * m_distort;
+    out(1) = in(1) * m_distort;
+
+
+  }
+  void Camera::distortEquiRefractive(const Eigen::Vector2d& in, Eigen::Vector2d& out) const{
+    Eigen::Vector2d temp;
+    distortRefractive(in, temp);
+    distortEquidist(temp, out);
+
+  }
+
+  void Camera::distortRefractive(const Eigen::Vector2d& in, Eigen::Vector2d& out, const double& refrac_index) const{
+    const double x2 = in(0) * in(0);
+    const double y2 = in(1) * in(1);
+    const double x_y = in(0) * in(1);
+    const double r2 = x2 + y2;
+    const double n = refrac_index;    // refractive index from state
+    const double n2 = n * n;
+
+    const double m_distort = n/sqrt(1 + r2 - (n2*r2));
+    out(0) = in(0) * m_distort;
+    out(1) = in(1) * m_distort;
+
+  }
+
+  void Camera::distortRefractive(const Eigen::Vector2d& in, Eigen::Vector2d& out, Eigen::Matrix2d& J) const{
+    const double x2 = in(0) * in(0);
+    const double y2 = in(1) * in(1);
+    const double x_y = in(0) * in(1);
+    const double r2 = x2 + y2;
+    const double n = refrac_ind_;
+    const double n2 = n * n;
+
+    const double m_distort = n/sqrt(1 + r2 - (n2*r2));
+    out(0) = in(0) * m_distort;
+    out(1) = in(1) * m_distort;
+
+    const double g = 1  + r2 - (n2*r2);
+
+    J(0,0) = n*pow(g, -2.0)*(sqrt(g)*x2*(n2 - 1) + pow(g, 1.5));
+    J(0,1) = n*pow(g, -1.5)*x_y*(n2 - 1);
+    J(1,0) = n*pow(g, -1.5)*x_y*(n2 - 1);
+    J(1,1) = n*pow(g, -2.0)*(sqrt(g)*y2*(n2 - 1) + pow(g, 1.5));
+  }
+
+  void Camera::distortEquiRefractive(const Eigen::Vector2d& in, Eigen::Vector2d& out, Eigen::Matrix2d& J) const{
+    Eigen::Vector2d temp;
+    Eigen::Matrix2d J_temp_refrac;
+    Eigen::Matrix2d J_temp_equi;
+
+    distortRefractive(in, temp, J_temp_refrac);
+    distortEquidist(temp, out, J_temp_equi);
+    J = J_temp_equi * J_temp_refrac;
+  }
+
+  void Camera::distortEquiRefractive(const Eigen::Vector2d& in, Eigen::Vector2d& out, Eigen::Matrix2d& J_equi, Eigen::Matrix2d& J_refrac) const{
+    Eigen::Vector2d temp;
+
+    distortRefractive(in, temp, J_refrac);
+    distortEquidist(temp, out, J_equi);
+  }
+
+  void Camera::distortRefractive(const Eigen::Vector2d& in, Eigen::Vector2d& out, const double& refrac_index ,Eigen::Matrix2d& J) const{
+    const double x2 = in(0) * in(0);
+    const double y2 = in(1) * in(1);
+    const double x_y = in(0) * in(1);
+    const double r2 = x2 + y2;
+    const double n = refrac_index; // refractive index from state
+    const double n2 = n * n;
+
+    const double m_distort = n/sqrt(1 + r2 - (n2*r2));
+    out(0) = in(0) * m_distort;
+    out(1) = in(1) * m_distort;
+
+    const double g = 1  + r2 - (n2*r2);
+
+    J(0,0) = n*pow(g, -2.0)*(sqrt(g)*x2*(n2 - 1) + pow(g, 1.5));
+    J(0,1) = n*pow(g, -1.5)*x_y*(n2 - 1);
+    J(1,0) = n*pow(g, -1.5)*x_y*(n2 - 1);
+    J(1,1) = n*pow(g, -2.0)*(sqrt(g)*y2*(n2 - 1) + pow(g, 1.5));
+  }
+
+  void Camera::distortEquiRefractive(const Eigen::Vector2d& in, Eigen::Vector2d& out, const double& refrac_index, Eigen::Matrix2d& J) const{
+    Eigen::Vector2d temp;
+    Eigen::Matrix2d J_temp_refrac;
+    Eigen::Matrix2d J_temp_equi;
+
+    distortRefractive(in, temp, refrac_index, J_temp_refrac);
+    distortEquidist(temp, out, J_temp_equi);
+    J = J_temp_equi * J_temp_refrac;
+  }
+
   void Camera::distortEquidist(const Eigen::Vector2d& in, Eigen::Vector2d& out) const{
     const double x2 = in(0) * in(0);
     const double y2 = in(1) * in(1);
@@ -228,9 +394,39 @@
       case RADTAN:
         distortRadtan(in,out);
         break;
+      case REFRAC:
+        distortRefractive(in,out);
+        break;
+      case EQUIDIST:
+        distortEquidist(in,out);
+        break;
+      case EQUIREFRAC:
+        distortEquiRefractive(in,out);
+        break;
+      case DS:
+        distortDoubleSphere(in,out);
+        break;
+      default:
+        distortRadtan(in,out);
+        break;
+    }
+  }
+
+  void Camera::distort(const Eigen::Vector2d& in, Eigen::Vector2d& out, const double& refrac_index) const{
+    std:: cout << "in distort with refrac_index" << std::endl;
+    switch(type_){
+      case RADTAN:
+        distortRadtan(in,out);
+        break;
+      case REFRAC:
+        distortRefractive(in,out,refrac_index);
+        break;
       case EQUIDIST:
         distortEquidist(in,out);
         break;
+      case EQUIREFRAC:
+        distortEquiRefractive(in,out);
+        break;
       case DS:
         distortDoubleSphere(in,out);
         break;
@@ -245,9 +441,39 @@
       case RADTAN:
         distortRadtan(in,out,J);
         break;
+      case REFRAC:
+        distortRefractive(in,out,J);
+        break;
+      case EQUIDIST:
+        distortEquidist(in,out,J);
+        break;
+      case EQUIREFRAC:
+        distortEquiRefractive(in,out,J);
+        break;
+      case DS:
+        distortDoubleSphere(in,out,J);
+        break;
+      default:
+        distortRadtan(in,out,J);
+        break;
+    }
+  }
+
+void Camera::distort(const Eigen::Vector2d& in, Eigen::Vector2d& out, const double& refrac_index, Eigen::Matrix2d& J) const{
+
+    switch(type_){
+      case RADTAN:
+        distortRadtan(in,out,J);
+        break;
+      case REFRAC:
+        distortRefractive(in,out, refrac_index,J);
+        break;
       case EQUIDIST:
         distortEquidist(in,out,J);
         break;
+      case EQUIREFRAC:
+        distortEquiRefractive(in,out, refrac_index,J);
+        break;
       case DS:
         distortDoubleSphere(in,out,J);
         break;
@@ -256,6 +482,7 @@
         break;
     }
   }
+  
 
   bool Camera::bearingToPixel(const Eigen::Vector3d& vec, cv::Point2f& c) const{
     // Project
@@ -272,6 +499,76 @@
     return true;
   }
 
+  bool Camera::bearingToPixel(const Eigen::Vector3d& vec, cv::Point2f& c, const double& refrac_index) const{
+    // Project
+    if(vec(2)<=0) return false;
+    const Eigen::Vector2d undistorted = Eigen::Vector2d(vec(0)/vec(2),vec(1)/vec(2));
+
+    // Distort
+    Eigen::Vector2d distorted;
+    // std::cout << "in bearingToPixel with refrac_index" << std::endl;
+    distort(undistorted,distorted, refrac_index);
+
+    // Shift origin and scale
+    c.x = static_cast<float>(K_(0, 0)*distorted(0) + K_(0, 2));
+    c.y = static_cast<float>(K_(1, 1)*distorted(1) + K_(1, 2));
+    return true;
+  }
+
+  bool Camera::bearingToPixel(const Eigen::Vector3d& vec, cv::Point2f& c, Eigen::Matrix<double,2,3>& J, Eigen::Matrix<double,2,1>& Jdpdn, const double& refrac_index) const{
+    // Project
+    if(vec(2)<=0) return false;
+    const Eigen::Vector2d undistorted = Eigen::Vector2d(vec(0)/vec(2),vec(1)/vec(2));
+    Eigen::Matrix<double,2,3> J1; J1.setZero();
+    J1(0,0) = 1.0/vec(2);
+    J1(0,2) = -vec(0)/pow(vec(2),2);
+    J1(1,1) = 1.0/vec(2);
+    J1(1,2) = -vec(1)/pow(vec(2),2);
+
+    // Distort
+    Eigen::Vector2d distorted;
+    Eigen::Matrix2d J2;
+    Eigen::Matrix2d J_equi;
+    Eigen::Matrix2d J_refrac;
+    // std::cout << "in bearingToPixel with distort(undistorted,distorted, refrac_index, J2)" << std::endl;
+    
+    // temporary fix for equirefractive
+    if (type_ == EQUIREFRAC){
+      distortEquiRefractive( undistorted,  distorted, J_equi, J_refrac);
+      J2 = J_equi * J_refrac;
+    }
+    else{
+      distort(undistorted,distorted, refrac_index, J2);
+    }
+
+    // Shift origin and scale
+    c.x = static_cast<float>(K_(0, 0)*distorted(0) + K_(0, 2));
+    c.y = static_cast<float>(K_(1, 1)*distorted(1) + K_(1, 2));
+    Eigen::Matrix2d J3; J3.setZero();
+    J3(0,0) = K_(0, 0);
+    J3(1,1) = K_(1, 1);
+
+    J = J3*J2*J1;
+    
+    // Jacobian of distorted point w.r.t. refractive index
+    double n = refrac_index;
+    double r2 = undistorted(0)*undistorted(0) + undistorted(1)*undistorted(1);
+    double g = 1 + r2 - (n*n*r2);
+    // clip g to avoid division by zero
+    // common_term = \frac{\g^{1/2}*n^2*r^2 + g^{3/2}}{g^2}
+    double common_term = (sqrt(g)*n*n*r2 + pow(g, 1.5))/(g*g);
+
+    // Jdpdn(0) = (K_(0, 0)*undistorted(0)*common_term);
+    // Jdpdn(1) = (K_(1, 1)*undistorted(1)*common_term);
+
+    Jdpdn(0) = undistorted(0)*common_term;
+    Jdpdn(1) = undistorted(1)*common_term;
+
+    Jdpdn = K_.block<2,2>(0,0)*J_equi*Jdpdn;
+
+    return true;
+  }
+
   bool Camera::bearingToPixel(const Eigen::Vector3d& vec, cv::Point2f& c, Eigen::Matrix<double,2,3>& J) const{
     // Project
     if(vec(2)<=0) return false;
@@ -285,6 +582,8 @@
     // Distort
     Eigen::Vector2d distorted;
     Eigen::Matrix2d J2;
+    // std::cout << "in bearingToPixel with distort(undistorted,distorted,J2)" << std::endl;
+
     distort(undistorted,distorted,J2);
 
     // Shift origin and scale
@@ -299,9 +598,42 @@
     return true;
   }
 
+  bool Camera::bearingToPixel(const Eigen::Vector3d& vec, cv::Point2f& c, Eigen::Matrix<double,2,3>& J, const double& refrac_index) const{
+    // Project
+    if(vec(2)<=0) return false;
+    const Eigen::Vector2d undistorted = Eigen::Vector2d(vec(0)/vec(2),vec(1)/vec(2));
+    Eigen::Matrix<double,2,3> J1; J1.setZero();
+    J1(0,0) = 1.0/vec(2);
+    J1(0,2) = -vec(0)/pow(vec(2),2);
+    J1(1,1) = 1.0/vec(2);
+    J1(1,2) = -vec(1)/pow(vec(2),2);
+
+    // Distort
+    Eigen::Vector2d distorted;
+    Eigen::Matrix2d J2;
+    // std::cout << "in bearingToPixel with distort(undistorted,distorted,refrac_index,J2)" << std::endl;
+
+    distort(undistorted,distorted,refrac_index,J2);
+
+    // Shift origin and scale
+    c.x = static_cast<float>(K_(0, 0)*distorted(0) + K_(0, 2));
+    c.y = static_cast<float>(K_(1, 1)*distorted(1) + K_(1, 2));
+    Eigen::Matrix2d J3; J3.setZero();
+    J3(0,0) = K_(0, 0);
+    J3(1,1) = K_(1, 1);
+
+    J = J3*J2*J1;
+
+    return true;
+  }
+
   bool Camera::bearingToPixel(const LWF::NormalVectorElement& n, cv::Point2f& c) const{
     return bearingToPixel(n.getVec(),c);
   }
+  
+  bool Camera::bearingToPixel(const LWF::NormalVectorElement& n, cv::Point2f& c, const double& refrac_index) const{
+    return bearingToPixel(n.getVec(),c, refrac_index);
+  }
 
   bool Camera::bearingToPixel(const LWF::NormalVectorElement& n, cv::Point2f& c, Eigen::Matrix<double,2,2>& J) const{
     Eigen::Matrix<double,3,2> J1;
@@ -312,6 +644,24 @@
     return success;
   }
 
+  bool Camera::bearingToPixel(const LWF::NormalVectorElement& n, cv::Point2f& c, Eigen::Matrix<double,2,2>& J, const double& refrac_index) const{
+    Eigen::Matrix<double,3,2> J1;
+    J1 = n.getM();
+    Eigen::Matrix<double,2,3> J2;
+    const bool success = bearingToPixel(n.getVec(),c,J2, refrac_index);
+    J = J2*J1;
+    return success;
+  }
+
+  bool Camera::bearingToPixel(const LWF::NormalVectorElement& n, cv::Point2f& c, Eigen::Matrix<double,2,2>& J, Eigen::Matrix<double,2,1>& Jdpdn, const double& refrac_index) const{
+    Eigen::Matrix<double,3,2> J1;
+    J1 = n.getM();
+    Eigen::Matrix<double,2,3> J2;
+    const bool success = bearingToPixel(n.getVec(),c,J2,Jdpdn, refrac_index);
+    J = J2*J1;
+    return success;
+  }
+
   bool Camera::pixelToBearing(const cv::Point2f& c,Eigen::Vector3d& vec) const{
     // Shift origin and scale
     Eigen::Vector2d y;
@@ -336,6 +686,12 @@
         success = true;
         break;
       }
+      // else{
+      //   std::cout << "Failed"<< std::endl;
+      //   std::cout << "e.dot(e) = " << e.dot(e) << std::endl;
+      //   std::cout << "ybar = " << ybar.transpose() << std::endl;
+        
+      // }
     }
     if(success){
       y = ybar;
@@ -344,13 +700,61 @@
     return success;
   }
 
+  bool Camera::pixelToBearingAnalytical(const cv::Point2f& c,Eigen::Vector3d& vec) const{
+    // Shift origin and scale
+    Eigen::Vector2d y;
+    // y(0) = (static_cast<double>(c.x) - K_(0, 2)) / K_(0, 0);
+    // y(1) = (static_cast<double>(c.y) - K_(1, 2)) / K_(1, 1);
+    y(0) = static_cast<double>(c.x);
+    y(1) = static_cast<double>(c.y);
+
+    
+    // Convert point to opencv format
+    cv::Mat y_mat(1, 2, CV_32F);
+    y_mat.at<float>(0, 0) = y(0);
+    y_mat.at<float>(0, 1) = y(1);
+    y_mat = y_mat.reshape(2); // Nx1, 2-channel
+
+
+    cv::Matx33d K(K_(0,0), 0, K_(0,2), 0, K_(1,1), K_(1,2), 0, 0, 1);
+    cv::Vec4d D(k1_, k2_, k3_, k4_);
+    // cv::undistortPoints(y_mat, undistorted, K, D);
+    cv::fisheye::undistortPoints(y_mat, y_mat, K, D);
+
+    y(0) = y_mat.at<float>(0, 0);
+    y(1) = y_mat.at<float>(0, 1);
+
+    // Undistort by analytical solution
+    const double x2 = y(0) * y(0);
+    const double y2 = y(1) * y(1);
+    const double r2 = x2 + y2;
+    const double n = refrac_ind_;
+    const double n2 = n * n;
+    const double m_undistort = sqrt((n2*r2) + n2 - r2);
+    y = y / m_undistort;
+
+    vec = Eigen::Vector3d(y(0),y(1),1.0).normalized();
+    // std::cout << "###### pixelToBearingAnalytical using refrac_ind_ = " << refrac_ind_ << "\n";
+    return true;
+  }
+
+
   bool Camera::pixelToBearing(const cv::Point2f& c,LWF::NormalVectorElement& n) const{
     Eigen::Vector3d vec;
-    bool success = pixelToBearing(c,vec);
+    bool success;
+    if (type_==REFRAC || type_==EQUIREFRAC){
+      success = pixelToBearingAnalytical(c,vec);
+    }
+    else{
+      success = pixelToBearing(c,vec);
+    }
+
     n.setFromVector(vec);
     return success;
   }
 
+
+
   void Camera::testCameraModel(){
     double d = 1e-4;
     LWF::NormalVectorElement b_s;
diff -urN rovio_ros2_ref/src/FeatureCoordinates.cpp reaqrovio_ros1_ref/src/FeatureCoordinates.cpp
--- rovio_ros2_ref/src/FeatureCoordinates.cpp	2026-07-24 19:53:08.747613608 +0300
+++ reaqrovio_ros1_ref/src/FeatureCoordinates.cpp	2026-07-24 19:52:58.993228056 +0300
@@ -1,3 +1,61 @@
+/*
+* For parts of code or files added at the Autonomous Robots Lab, Norwegian University of Science and Technology, the following license is applicable:
+*
+* Copyright (c) 2024, Autonomous Robots Lab, Norwegian University of Science and Technology
+*
+* Redistribution and use in source and binary forms, with or without
+* modification, are permitted provided that the following conditions are met:
+*
+* 1. Redistributions of source code must retain the above copyright notice, this
+*    list of conditions and the following disclaimer.
+*
+* 2. Redistributions in binary form must reproduce the above copyright notice,
+*    this list of conditions and the following disclaimer in the documentation
+*    and/or other materials provided with the distribution.
+*
+* 3. Neither the name of the copyright holder nor the names of its
+*    contributors may be used to endorse or promote products derived from
+*    this software without specific prior written permission.
+*
+* THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+* AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+* IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+* DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+* FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+* DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+* SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+* CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+* OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+* OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*
+* For parts of code or files directly adopted from the Autonomous Systems Lab, the following license is applicable (as from original repository for ROVIO: https://github.com/ethz-asl/rovio):
+*
+* Copyright (c) 2014, Autonomous Systems Lab
+* All rights reserved.
+*
+* Redistribution and use in source and binary forms, with or without
+* modification, are permitted provided that the following conditions are met:
+* * Redistributions of source code must retain the above copyright
+* notice, this list of conditions and the following disclaimer.
+* * Redistributions in binary form must reproduce the above copyright
+* notice, this list of conditions and the following disclaimer in the
+* documentation and/or other materials provided with the distribution.
+* * Neither the name of the Autonomous Systems Lab, ETH Zurich nor the
+* names of its contributors may be used to endorse or promote products
+* derived from this software without specific prior written permission.
+*
+* THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+* ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+* WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+* DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+* FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+* DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+* SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+* CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+* OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+* OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*/
+
 #include "rovio/FeatureCoordinates.hpp"
 
 namespace rovio{
@@ -50,7 +108,7 @@
 
   const cv::Point2f& FeatureCoordinates::get_c() const{
     if(!com_c()){
-      std::cout << "    \033[31mERROR: No valid coordinate data!\033[0m" << std::endl;
+      std::cout << "    \033[31mERROR: No valid coordinate data [1]!\033[0m" << std::endl;
     }
     return c_;
   }
@@ -67,7 +125,7 @@
 
   const LWF::NormalVectorElement& FeatureCoordinates::get_nor() const{
     if(!com_nor()){
-      std::cout << "    \033[31mERROR: No valid coordinate data!\033[0m" << std::endl;
+      std::cout << "    \033[31mERROR: No valid coordinate data [2]!\033[0m" << std::endl;
     }
     return nor_;
   }
@@ -76,7 +134,7 @@
     assert(mpCamera_ != nullptr);
     if(!mpCamera_->bearingToPixel(get_nor(),c_,matrix2dTemp_)){
       matrix2dTemp_.setZero();
-      std::cout << "    \033[31mERROR: No valid coordinate data!\033[0m" << std::endl;
+      std::cout << "    \033[31mERROR: No valid coordinate data [3]!\033[0m" << std::endl;
     }
     return matrix2dTemp_;
   }
@@ -91,10 +149,12 @@
     }
   }
 
+
   void FeatureCoordinates::set_nor(const LWF::NormalVectorElement& nor, const bool resetWarp){
     nor_ = nor;
     valid_nor_ = true;
     valid_c_ = false;
+
     if(trackWarping_ && resetWarp){
       valid_warp_c_ = false;
       valid_warp_nor_ = false;
diff -urN rovio_ros2_ref/src/feature_tracker_node.cpp reaqrovio_ros1_ref/src/feature_tracker_node.cpp
--- rovio_ros2_ref/src/feature_tracker_node.cpp	2026-07-24 19:53:08.747950814 +0300
+++ reaqrovio_ros1_ref/src/feature_tracker_node.cpp	2026-07-24 19:52:58.993923458 +0300
@@ -29,8 +29,9 @@
 #include "../include/rovio/featureTracker.hpp"
 
 int main(int argc, char** argv) {
-  rclcpp::init(argc, argv);
-  auto featureTrackerNode= std::make_shared<rovio::FeatureTrackerNode>();
-  rclcpp::spin(featureTrackerNode);
+  ros::init(argc, argv, "FeatureTrackerNode");
+  ros::NodeHandle nh;
+  rovio::FeatureTrackerNode featureTrackerNode(nh);
+  ros::spin();
   return 0;
 }
diff -urN rovio_ros2_ref/src/rovio_node.cpp reaqrovio_ros1_ref/src/rovio_node.cpp
--- rovio_ros2_ref/src/rovio_node.cpp	2026-07-24 19:53:08.747950814 +0300
+++ reaqrovio_ros1_ref/src/rovio_node.cpp	2026-07-24 19:52:58.993923458 +0300
@@ -1,4 +1,35 @@
 /*
+* For parts of code or files added at the Autonomous Robots Lab, Norwegian University of Science and Technology, the following license is applicable:
+*
+* Copyright (c) 2024, Autonomous Robots Lab, Norwegian University of Science and Technology
+*
+* Redistribution and use in source and binary forms, with or without
+* modification, are permitted provided that the following conditions are met:
+*
+* 1. Redistributions of source code must retain the above copyright notice, this
+*    list of conditions and the following disclaimer.
+*
+* 2. Redistributions in binary form must reproduce the above copyright notice,
+*    this list of conditions and the following disclaimer in the documentation
+*    and/or other materials provided with the distribution.
+*
+* 3. Neither the name of the copyright holder nor the names of its
+*    contributors may be used to endorse or promote products derived from
+*    this software without specific prior written permission.
+*
+* THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+* AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+* IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+* DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+* FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+* DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+* SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+* CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+* OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+* OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*
+* For parts of code or files directly adopted from the Autonomous Systems Lab, the following license is applicable (as from original repository for ROVIO: https://github.com/ethz-asl/rovio):
+*
 * Copyright (c) 2014, Autonomous Systems Lab
 * All rights reserved.
 *
@@ -23,7 +54,6 @@
 * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
 * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
-*
 */
 
 
@@ -33,8 +63,9 @@
 
 #pragma GCC diagnostic push
 #pragma GCC diagnostic ignored "-Wunused-parameter"
-#include <rclcpp/rclcpp.hpp>
-#include <geometry_msgs/msg/pose_stamped.hpp>
+#include <ros/ros.h>
+#include <ros/package.h>
+#include <geometry_msgs/Pose.h>
 #pragma GCC diagnostic pop
 
 #include "rovio/RovioFilter.hpp"
@@ -75,57 +106,61 @@
 
 typedef rovio::RovioFilter<rovio::FilterState<nMax_,nLevels_,patchSize_,nCam_,nPose_>> mtFilter;
 
+#ifdef MAKE_SCENE
+static rovio::RovioScene<mtFilter> mRovioScene;
 
+void idleFunc(){
+  ros::spinOnce();
+  mRovioScene.drawScene(mRovioScene.mpFilter_->safe_);
+}
+#endif
+
+int main(int argc, char** argv){
+  ros::init(argc, argv, "rovio");
+  ros::NodeHandle nh;
+  ros::NodeHandle nh_private("~");
 
-/**
- * @brief Function to read the camera calibration parameters. File path are ros params of node.
- * @param mpFilter
- * @param node
- * @return None
- */
-void readCameraConfig(std::shared_ptr<mtFilter> mpFilter,
-                      std::shared_ptr<rovio::RovioNode<mtFilter>> node) {
+  std::string rootdir = ros::package::getPath("rovio"); // Leaks memory
+  std::string filter_config = rootdir + "/cfg/rovio.info";
+
+  nh_private.param("filter_config", filter_config, filter_config);
+
+  // Filter
+  std::shared_ptr<mtFilter> mpFilter(new mtFilter);
+  mpFilter->readFromInfo(filter_config);
+
+  // Force the camera calibration paths to the ones from ROS parameters.
   for (unsigned int camID = 0; camID < nCam_; ++camID) {
     std::string camera_config;
-    if (node->get_parameter("camera" + std::to_string(camID)
+    if (nh_private.getParam("camera" + std::to_string(camID)
                             + "_config", camera_config)) {
-      std::cout << "Camera config: " << camera_config << std::endl;
       mpFilter->cameraCalibrationFile_[camID] = camera_config;
     }
   }
-}
+  mpFilter->refreshProperties();
 
-/**
- * @brief Function to declare the camera config parameters
- * @param ROVIO node share ptr
- * @return none
- */
-void declareParameters(std::shared_ptr<rovio::RovioNode<mtFilter>> node)
-{
-  for (unsigned int camID = 0; camID < nCam_; ++camID) {
-    std::string camera_config;
-    node->declare_parameter("camera" + std::to_string(camID)
-                            + "_config","");
-  }
-  node->declare_parameter("filter_config", "");
-}
+  // Set refractive index from ROS parameter. CUATION: This overwrites the value in rovio.info file.
+  double refractive_index;
+  if (nh_private.getParam("refractive_index", refractive_index)) {
 
-int main(int argc, char** argv){
-  rclcpp::init(argc, argv);
-  // Filter
-  std::shared_ptr<mtFilter> mpFilter(new mtFilter);
-  std::string filter_config;
+    ROS_WARN("Setting refractive index to %f from ROS parameter.", refractive_index);
+    mpFilter->setRefractiveIndex(refractive_index);
+  }
 
   // Node
-  std::shared_ptr<rovio::RovioNode<rovio::RovioFilter<rovio::FilterState<25, 4, 6, 1, 0>>>> node;
-  node = std::make_shared<rovio::RovioNode<mtFilter>>(mpFilter);
-  declareParameters(node);
-  node->get_parameter("filter_config", filter_config);
-  mpFilter->readFromInfo(filter_config);
-  std::cout << "Filter config: " << filter_config << std::endl;
-  readCameraConfig(mpFilter, node);
-  mpFilter->refreshProperties();
-  node->makeTest();
-  rclcpp::spin(node);
+  rovio::RovioNode<mtFilter> rovioNode(nh, nh_private, mpFilter);
+  rovioNode.makeTest();
+
+#ifdef MAKE_SCENE
+  // Scene
+  std::string mVSFileName = rootdir + "/shaders/shader.vs";
+  std::string mFSFileName = rootdir + "/shaders/shader.fs";
+  mRovioScene.initScene(argc,argv,mVSFileName,mFSFileName,mpFilter);
+  mRovioScene.setIdleFunction(idleFunc);
+  mRovioScene.addKeyboardCB('r',[&rovioNode]() mutable {rovioNode.requestReset();});
+  glutMainLoop();
+#else
+  ros::spin();
+#endif
   return 0;
 }
diff -urN rovio_ros2_ref/src/rovio_rosbag_loader.cpp reaqrovio_ros1_ref/src/rovio_rosbag_loader.cpp
--- rovio_ros2_ref/src/rovio_rosbag_loader.cpp	2026-07-24 19:53:08.747950814 +0300
+++ reaqrovio_ros1_ref/src/rovio_rosbag_loader.cpp	2026-07-24 19:52:58.993923458 +0300
@@ -26,11 +26,9 @@
 *
 */
 
-#include <rclcpp/rclcpp.hpp>
-#include <rclcpp/serialization.hpp>
-#include <rosbag2_cpp/reader.hpp>
-#include <rosbag2_cpp/writer.hpp>
-#include <rosbag2_storage/storage_options.hpp>
+#include <ros/package.h>
+#include <rosbag/bag.h>
+#include <rosbag/view.h>
 #include <memory>
 #include <iostream>
 #include <locale>
@@ -41,10 +39,6 @@
 #include <boost/foreach.hpp>
 #include <boost/date_time/posix_time/posix_time.hpp>
 #include <boost/date_time/posix_time/posix_time_io.hpp>
-
-#include "rovio/RovioNode.hpp"
-#include "geometry_msgs/msg/point_stamped.hpp"
-#include "geometry_msgs/msg/pose_stamped.hpp"
 #define foreach BOOST_FOREACH
 
 #ifdef ROVIO_NMAXFEATURE
@@ -79,144 +73,60 @@
 
 typedef rovio::RovioFilter<rovio::FilterState<nMax_,nLevels_,patchSize_,nCam_,nPose_>> mtFilter;
 
-/**
- * @brief function to deserialize a template sample message, to read from a rosbag.
- * @tparam T
- * @param msg
- * @return T type. Msg type/
- */
-template <typename T>
-T deserializeMessage(auto msg) {
-  rclcpp::SerializedMessage extractedMsg(*msg->serialized_data);
-  rclcpp::Serialization<T> deserializer;
-  T returnMsg;
-  deserializer.deserialize_message(&extractedMsg, &returnMsg);
-  return returnMsg;
-}
-
-
-/**
- *
- * @brief function to serialize the message, to store in the bag file.
- * @tparam Message type
- * @param msg to serialize
- * @return generalized serialized msg to store in bag file
- */
-template <typename T>
-rclcpp::SerializedMessage serializeMessage(T msgToSerialize ) {
-  rclcpp::SerializedMessage serializedMsg;
-  rclcpp::Serialization<T> serializer;
-  serializer.serialize_message(&msgToSerialize, &serializedMsg);
-  return serializedMsg;
-}
-
-
-/**
- * @brief Function to convert the Point stamped message to the pose stamped message
- * @param PointStamped message reference
- * @param PoseStamped message reference
- * @return None
- */
-void pointToPose(const geometry_msgs::msg::PointStamped pointMsg,
-                geometry_msgs::msg::PoseStamped::SharedPtr poseMsg) {
-  poseMsg->header.stamp = pointMsg.header.stamp;
-  poseMsg->header.frame_id = pointMsg.header.frame_id;
-  poseMsg->pose.position.x = pointMsg.point.x;
-  poseMsg->pose.position.y = pointMsg.point.y;
-  poseMsg->pose.position.z = pointMsg.point.z;
-  poseMsg->pose.orientation.x = 0;
-  poseMsg->pose.orientation.y = 0;
-  poseMsg->pose.orientation.z = 0;
-  poseMsg->pose.orientation.w = 1;
-}
-
-
-/**
- * @brief Function to declare the camera config parameters
- * @param ROVIO node share ptr
- * @return none
- */
-void declareParameters(std::shared_ptr<rovio::RovioNode<mtFilter>> node) {
-  for (unsigned int camID = 0; camID < nCam_; ++camID) {
-    std::string camera_config;
-    node->declare_parameter("camera" + std::to_string(camID)
-                            + "_config","");
-  }
-  node->declare_parameter("filter_config", "");
-  node->declare_parameter("record_odometry", true);
-  node->declare_parameter("record_pose_with_covariance_stamped", true);
-  node->declare_parameter("record_transform", true);
-  node->declare_parameter("record_extrinsics", true);
-  node->declare_parameter("record_imu_bias", true);
-  node->declare_parameter("record_pcl", true);
-  node->declare_parameter("record_markers", true);
-  node->declare_parameter("record_patch", false);
-  node->declare_parameter("reset_trigger", 0.0);
-}
-
-
-/**
- * @brief Function to read the camera calibration parameters. File path are ros params of node.
- * @param mpFilter
- * @param node
- * @return None
- */
-void readCameraConfig(std::shared_ptr<mtFilter> mpFilter,
-                      std::shared_ptr<rovio::RovioNode<mtFilter>> node) {
-  for (unsigned int camID = 0; camID < nCam_; ++camID) {
-    std::string camera_config;
-    if (node->get_parameter("camera" + std::to_string(camID)
-                            + "_config", camera_config)) {
-      mpFilter->cameraCalibrationFile_[camID] = camera_config;
-                            }
-  }
-}
+int main(int argc, char** argv){
+  ros::init(argc, argv, "rovio");
+  ros::NodeHandle nh;
+  ros::NodeHandle nh_private("~");
 
+  std::string rootdir = ros::package::getPath("rovio"); // Leaks memory
+  std::string filter_config = rootdir + "/cfg/rovio.info";
 
-int main(int argc, char** argv){
-  rclcpp::init(argc, argv);
-  std::string filter_config;
+  nh_private.param("filter_config", filter_config, filter_config);
 
   // Filter
   std::shared_ptr<mtFilter> mpFilter(new mtFilter);
-  auto rovioNode = std::make_shared<rovio::RovioNode<mtFilter>>(mpFilter);
-  declareParameters(rovioNode);
-  rovioNode->get_parameter("filter_config", filter_config);
   mpFilter->readFromInfo(filter_config);
 
   // Force the camera calibration paths to the ones from ROS parameters.
-  readCameraConfig(mpFilter, rovioNode);
+  for (unsigned int camID = 0; camID < nCam_; ++camID) {
+    std::string camera_config;
+    if (nh_private.getParam("camera" + std::to_string(camID)
+                            + "_config", camera_config)) {
+      mpFilter->cameraCalibrationFile_[camID] = camera_config;
+    }
+  }
   mpFilter->refreshProperties();
-  
-  rovioNode->makeTest();
+
+  // Node
+  rovio::RovioNode<mtFilter> rovioNode(nh, nh_private, mpFilter);
+  rovioNode.makeTest();
   double resetTrigger = 0.0;
-  rovioNode->get_parameter("record_odometry", rovioNode->forceOdometryPublishing_);
-  rovioNode->get_parameter("record_pose_with_covariance_stamped", rovioNode->forcePoseWithCovariancePublishing_);
-  rovioNode->get_parameter("record_transform", rovioNode->forceTransformPublishing_);
-  rovioNode->get_parameter("record_extrinsics", rovioNode->forceExtrinsicsPublishing_);
-  rovioNode->get_parameter("record_imu_bias", rovioNode->forceImuBiasPublishing_);
-  rovioNode->get_parameter("record_pcl", rovioNode->forcePclPublishing_);
-  rovioNode->get_parameter("record_markers", rovioNode->forceMarkersPublishing_);
-  rovioNode->get_parameter("record_patch", rovioNode->forcePatchPublishing_);
-  rovioNode->get_parameter("reset_trigger", resetTrigger);
+  nh_private.param("record_odometry", rovioNode.forceOdometryPublishing_, rovioNode.forceOdometryPublishing_);
+  nh_private.param("record_pose_with_covariance_stamped", rovioNode.forcePoseWithCovariancePublishing_, rovioNode.forcePoseWithCovariancePublishing_);
+  nh_private.param("record_transform", rovioNode.forceTransformPublishing_, rovioNode.forceTransformPublishing_);
+  nh_private.param("record_extrinsics", rovioNode.forceExtrinsicsPublishing_, rovioNode.forceExtrinsicsPublishing_);
+  nh_private.param("record_imu_bias", rovioNode.forceImuBiasPublishing_, rovioNode.forceImuBiasPublishing_);
+  nh_private.param("record_pcl", rovioNode.forcePclPublishing_, rovioNode.forcePclPublishing_);
+  nh_private.param("record_markers", rovioNode.forceMarkersPublishing_, rovioNode.forceMarkersPublishing_);
+  nh_private.param("record_patch", rovioNode.forcePatchPublishing_, rovioNode.forcePatchPublishing_);
+  nh_private.param("reset_trigger", resetTrigger, resetTrigger);
 
   std::cout << "Recording";
-  if(rovioNode->forceOdometryPublishing_) std::cout << ", odometry";
-  if(rovioNode->forceTransformPublishing_) std::cout << ", transform";
-  if(rovioNode->forceExtrinsicsPublishing_) std::cout << ", extrinsics";
-  if(rovioNode->forceImuBiasPublishing_) std::cout << ", imu biases";
-  if(rovioNode->forcePclPublishing_) std::cout << ", point cloud";
-  if(rovioNode->forceMarkersPublishing_) std::cout << ", markers";
-  if(rovioNode->forcePatchPublishing_) std::cout << ", patch data";
+  if(rovioNode.forceOdometryPublishing_) std::cout << ", odometry";
+  if(rovioNode.forceTransformPublishing_) std::cout << ", transform";
+  if(rovioNode.forceExtrinsicsPublishing_) std::cout << ", extrinsics";
+  if(rovioNode.forceImuBiasPublishing_) std::cout << ", imu biases";
+  if(rovioNode.forcePclPublishing_) std::cout << ", point cloud";
+  if(rovioNode.forceMarkersPublishing_) std::cout << ", markers";
+  if(rovioNode.forcePatchPublishing_) std::cout << ", patch data";
   std::cout << std::endl;
 
-  rosbag2_cpp::Reader bagIn;
+  rosbag::Bag bagIn;
   std::string rosbag_filename = "dataset.bag";
-  rovioNode->declare_parameter("rosbag_filename", rosbag_filename);
-  rovioNode->get_parameter("rosbag_filename", rosbag_filename);
-  bagIn.open(rosbag_filename);
+  nh_private.param("rosbag_filename", rosbag_filename, rosbag_filename);
+  bagIn.open(rosbag_filename, rosbag::bagmode::Read);
 
-  rosbag2_cpp::Writer bagOut;
+  rosbag::Bag bagOut;
   std::size_t found = rosbag_filename.find_last_of("/");
   std::string file_path = rosbag_filename.substr(0,found);
   std::string file_name = rosbag_filename.substr(found+1);
@@ -229,114 +139,82 @@
   boost::posix_time::time_facet* facet = new boost::posix_time::time_facet();
   facet->format("%Y-%m-%d-%H-%M-%S");
   stream.imbue(std::locale(std::locale::classic(), facet));
-  stream << rovioNode->get_clock()->now().seconds() << "_" << nMax_ << "_" << nLevels_ << "_" << patchSize_ << "_" << nCam_  << "_" << nPose_;
+  stream << ros::Time::now().toBoost() << "_" << nMax_ << "_" << nLevels_ << "_" << patchSize_ << "_" << nCam_  << "_" << nPose_;
   std::string filename_out = file_path + "/rovio/" + stream.str();
-  rovioNode->declare_parameter("filename_out", filename_out);
-  rovioNode->get_parameter("filename_out", filename_out);
+  nh_private.param("filename_out", filename_out, filename_out);
   std::string rosbag_filename_out = filename_out + ".bag";
   std::string info_filename_out = filename_out + ".info";
-  std::string gt_topic_name = "/gt";
-  rovioNode->declare_parameter("gt_topic_name", gt_topic_name);
-  rovioNode->get_parameter("gt_topic_name",gt_topic_name);
-  std::cout << "GT topic: " << gt_topic_name << std::endl;
-  bagOut.open(rosbag_filename_out);
+  std::cout << "Storing output to: " << rosbag_filename_out << std::endl;
+  bagOut.open(rosbag_filename_out, rosbag::bagmode::Write);
 
   // Copy info
   std::ifstream  src(filter_config, std::ios::binary);
   std::ofstream  dst(info_filename_out,   std::ios::binary);
   dst << src.rdbuf();
+
   std::vector<std::string> topics;
   std::string imu_topic_name = "/imu0";
-  rovioNode->declare_parameter("imu_topic_name", imu_topic_name);
-  rovioNode->get_parameter("imu_topic_name", imu_topic_name);
+  nh_private.param("imu_topic_name", imu_topic_name, imu_topic_name);
   std::string cam0_topic_name = "/cam0/image_raw";
-  rovioNode->declare_parameter("cam0_topic_name", cam0_topic_name);
-  rovioNode->get_parameter("cam0_topic_name", cam0_topic_name);
+  nh_private.param("cam0_topic_name", cam0_topic_name, cam0_topic_name);
   std::string cam1_topic_name = "/cam1/image_raw";
-  rovioNode->declare_parameter("cam1_topic_name", cam1_topic_name);
-  rovioNode->get_parameter("cam1_topic_name", cam1_topic_name);
-  std::string odometry_topic_name = rovioNode->pubOdometry_->get_topic_name();
-  //std::string transform_topic_name = rovioNode->pubTransform_->get_topic_name();
+  nh_private.param("cam1_topic_name", cam1_topic_name, cam1_topic_name);
+  std::string odometry_topic_name = rovioNode.pubOdometry_.getTopic();
+  std::string transform_topic_name = rovioNode.pubTransform_.getTopic();
   std::string extrinsics_topic_name[mtFilter::mtState::nCam_];
   for(int camID=0;camID<mtFilter::mtState::nCam_;camID++){
-    extrinsics_topic_name[camID] = rovioNode->pubExtrinsics_[camID]->get_topic_name();
+    extrinsics_topic_name[camID] = rovioNode.pubExtrinsics_[camID].getTopic();
   }
-  std::string imu_bias_topic_name = rovioNode->pubImuBias_->get_topic_name();
-  std::string pcl_topic_name = rovioNode->pubPcl_->get_topic_name();
-  std::string u_rays_topic_name = rovioNode->pubMarkers_->get_topic_name();
-  std::string patch_topic_name = rovioNode->pubPatch_->get_topic_name();
-  std::string health_topic_name = rovioNode->healthMonitorPublisher->get_topic_name();
+  std::string imu_bias_topic_name = rovioNode.pubImuBias_.getTopic();
+  std::string pcl_topic_name = rovioNode.pubPcl_.getTopic();
+  std::string u_rays_topic_name = rovioNode.pubMarkers_.getTopic();
+  std::string patch_topic_name = rovioNode.pubPatch_.getTopic();
 
   topics.push_back(std::string(imu_topic_name));
   topics.push_back(std::string(cam0_topic_name));
   topics.push_back(std::string(cam1_topic_name));
+  rosbag::View view(bagIn, rosbag::TopicQuery(topics));
 
 
   bool isTriggerInitialized = false;
   double lastTriggerTime = 0.0;
-  while (bagIn.has_next()) {
-    auto serializedMsg = bagIn.read_next();
-    if(serializedMsg->topic_name == imu_topic_name){
-      sensor_msgs::msg::Imu imuMsg = deserializeMessage<sensor_msgs::msg::Imu>(serializedMsg);
-      sensor_msgs::msg::Imu::ConstPtr imuMsgPtr = std::make_shared<sensor_msgs::msg::Imu>(imuMsg);
-      if (imuMsgPtr != NULL) rovioNode->imuCallback(imuMsgPtr);
+  for(rosbag::View::iterator it = view.begin();it != view.end() && ros::ok();it++){
+    if(it->getTopic() == imu_topic_name){
+      sensor_msgs::Imu::ConstPtr imuMsg = it->instantiate<sensor_msgs::Imu>();
+      if (imuMsg != NULL) rovioNode.imuCallback(imuMsg);
     }
-    if(serializedMsg->topic_name == cam0_topic_name){
-      sensor_msgs::msg::Image imgMsg =  deserializeMessage<sensor_msgs::msg::Image>(serializedMsg);
-      sensor_msgs::msg::Image::ConstPtr imgMsgPtr = std::make_shared<sensor_msgs::msg::Image>(imgMsg);
-      if (imgMsgPtr != NULL) rovioNode->imgCallback0(imgMsgPtr);
+    if(it->getTopic() == cam0_topic_name){
+      sensor_msgs::ImageConstPtr imgMsg = it->instantiate<sensor_msgs::Image>();
+      if (imgMsg != NULL) rovioNode.imgCallback0(imgMsg);
     }
-    if(serializedMsg->topic_name == cam1_topic_name){
-      sensor_msgs::msg::Image imgMsg2 = deserializeMessage<sensor_msgs::msg::Image>(serializedMsg);
-      sensor_msgs::msg::Image::ConstPtr imgMsg2Ptr = std::make_shared<sensor_msgs::msg::Image>(imgMsg2);
-      if (imgMsg2Ptr != NULL) rovioNode->imgCallback1(imgMsg2Ptr);
+    if(it->getTopic() == cam1_topic_name){
+      sensor_msgs::ImageConstPtr imgMsg = it->instantiate<sensor_msgs::Image>();
+      if (imgMsg != NULL) rovioNode.imgCallback1(imgMsg);
     }
-	if(serializedMsg->topic_name == gt_topic_name) {
-		geometry_msgs::msg::PointStamped gtPose = deserializeMessage<geometry_msgs::msg::PointStamped>(serializedMsg);
-    geometry_msgs::msg::PoseStamped::SharedPtr poseMsgPtr = std::make_shared<geometry_msgs::msg::PoseStamped>();
-	  pointToPose(gtPose, poseMsgPtr);
-	  bagOut.write(*poseMsgPtr, gt_topic_name, rovioNode->get_clock()->now());
-	}
-    rclcpp::spin_some(rovioNode);
-
-    if(rovioNode->gotFirstMessages_){
-      static double lastSafeTime = rovioNode->mpFilter_->safe_.t_;
-      if(rovioNode->mpFilter_->safe_.t_ > lastSafeTime) {
-        if(rovioNode->forceOdometryPublishing_)
-        {
-          bagOut.write(rovioNode->odometryMsg_, odometry_topic_name, rovioNode->get_clock()->now());
-        }
-        //if(rovioNode->forceTransformPublishing_) bagOut.write(transform_topic_name,rovioNode->get_clock()->now(),rovioNode->transformMsg_);
-        for(int camID=0;camID<mtFilter::mtState::nCam_;camID++)
-        {
-          if(rovioNode->forceExtrinsicsPublishing_) {
-            bagOut.write(rovioNode->extrinsicsMsg_[camID],
-              extrinsics_topic_name[camID],rovioNode->get_clock()->now());
-          }
-        }
-        if(rovioNode->forceImuBiasPublishing_) {
-          bagOut.write(rovioNode->imuBiasMsg_,imu_bias_topic_name,rovioNode->get_clock()->now());
-        }
-        if(rovioNode->forcePclPublishing_) {
-          bagOut.write(rovioNode->pclMsg_, pcl_topic_name, rovioNode->get_clock()->now());
-        }
-        if(rovioNode->forceMarkersPublishing_) {
-          bagOut.write(rovioNode->odometryMsg_,odometry_topic_name,rovioNode->get_clock()->now());
-        }
-        if(rovioNode->forcePatchPublishing_) {
-          bagOut.write(rovioNode->patchMsg_, patch_topic_name, rovioNode->get_clock()->now());
-        }
-        bagOut.write(rovioNode->healthMonitorMsg, health_topic_name, rovioNode->get_clock()->now());
-        lastSafeTime = rovioNode->mpFilter_->safe_.t_;
+    ros::spinOnce();
+
+    if(rovioNode.gotFirstMessages_){
+      static double lastSafeTime = rovioNode.mpFilter_->safe_.t_;
+      if(rovioNode.mpFilter_->safe_.t_ > lastSafeTime){
+        if(rovioNode.forceOdometryPublishing_) bagOut.write(odometry_topic_name,ros::Time::now(),rovioNode.odometryMsg_);
+        if(rovioNode.forceTransformPublishing_) bagOut.write(transform_topic_name,ros::Time::now(),rovioNode.transformMsg_);
+        for(int camID=0;camID<mtFilter::mtState::nCam_;camID++){
+          if(rovioNode.forceExtrinsicsPublishing_) bagOut.write(extrinsics_topic_name[camID],ros::Time::now(),rovioNode.extrinsicsMsg_[camID]);
+        }
+        if(rovioNode.forceImuBiasPublishing_) bagOut.write(imu_bias_topic_name,ros::Time::now(),rovioNode.imuBiasMsg_);
+        if(rovioNode.forcePclPublishing_) bagOut.write(pcl_topic_name,ros::Time::now(),rovioNode.pclMsg_);
+        if(rovioNode.forceMarkersPublishing_) bagOut.write(u_rays_topic_name,ros::Time::now(),rovioNode.markerMsg_);
+        if(rovioNode.forcePatchPublishing_) bagOut.write(patch_topic_name,ros::Time::now(),rovioNode.patchMsg_);
+        lastSafeTime = rovioNode.mpFilter_->safe_.t_;
       }
       if(!isTriggerInitialized){
         lastTriggerTime = lastSafeTime;
         isTriggerInitialized = true;
       }
       if(resetTrigger>0.0 && lastSafeTime - lastTriggerTime > resetTrigger){
-        rovioNode->requestReset();
-        rovioNode->mpFilter_->init_.state_.WrWM() = rovioNode->mpFilter_->safe_.state_.WrWM();
-        rovioNode->mpFilter_->init_.state_.qWM() = rovioNode->mpFilter_->safe_.state_.qWM();
+        rovioNode.requestReset();
+        rovioNode.mpFilter_->init_.state_.WrWM() = rovioNode.mpFilter_->safe_.state_.WrWM();
+        rovioNode.mpFilter_->init_.state_.qWM() = rovioNode.mpFilter_->safe_.state_.qWM();
         lastTriggerTime = lastSafeTime;
       }
     }
